migrations/deleteDuplicateRecordsAcrossCollections/deleteDuplicateRecords.js

```diff
@@ -0,0 +1,171 @@
+/**
+ * name : deleteDuplicateRecords.js
+ * author : Saish Borkar
+ * created-date : 1 Jul 2024
+ * Description : Delete duplicate records having same userId and solutionId to enforce compound key 
+ *               and data integrity
+ */
+
+const path = require("path");
+let rootPath = path.join(__dirname, "../../");
+require("dotenv").config({ path: rootPath + "/.env" });
+const {validate : uuidValidate,v4 : uuidV4} = require('uuid');
+let _ = require("lodash");
+let mongoUrl = process.env.MONGODB_URL;
+let dbName = mongoUrl.split("/").pop();
+let url = mongoUrl.split(dbName)[0];
+var MongoClient = require("mongodb").MongoClient;
+
+var fs = require("fs");
+
+function generateUUId() {
+    return uuidV4();
+  }
+  
+
+(async () => {
+  let connection = await MongoClient.connect(url, { useNewUrlParser: true });
+  let db = connection.db(dbName);
+
+  try {
+    let pipeline = [
+      {
+        $match: {
+          userId: { $exists: true, $ne: null },
+          solutionId: { $exists: true, $ne: null },
+        },
+      },
+      {
+        $group: {
+          _id: { solutionId: "$solutionId", userId: "$userId" },
+          count: { $sum: 1 },
+          docs: { $push: "$$ROOT" },
+        },
+      },
+      {
+        $match: {
+          count: { $gt: 1 },
+        },
+      },
+      {
+        $project: {
+          _id: 0,
+          duplicateCount: "$count",
+          duplicateArray: "$docs",
+        },
+      },
+    ];
+    let duplicateArray = await db
+      .collection("projects")
+      .aggregate(pipeline)
+      .toArray();
+
+    let successfullyDeletedRecords = [];
+    let failedToDeletedRecords = [];
+    for (let duplicateRecord of duplicateArray) {
+      try {
+
+        let duplicateArray = duplicateRecord.duplicateArray;
+        let completedProjects = [];
+        let inCompleteProjects = [];
+        let inProgressProjects = [];
+        let toBeDeletedRecords = [];
+
+        for (let i = 0; i < duplicateArray.length; i++) {
+          if (duplicateArray[i].status == "submitted") {
+            completedProjects.push(duplicateArray[i]);
+          } else if (duplicateArray[i].status == "started") {
+            inCompleteProjects.push(duplicateArray[i]);
+          }else if(duplicateArray[i].status == 'inProgress'){
+            inProgressProjects.push(duplicateArray[i])
+          }
+        }
+
+        completedProjects.sort(sortByCreatedAtDescending);
```

Comment:
dont check only createdAt if project has duplicate then check which one is in started and which one is in inprogress then delete the started One keep the inprogress on @borkarsaish65

Reply:
I have used this exact logic. Please check below code. Here I am just sorting all the records in descending order.